components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -284,53 +295,99 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
         }
         int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
-        String[] resultedUserList = ArrayUtils.EMPTY_STRING_ARRAY;
-        for (String key : claims.keySet()) {
-            String value = claims.get(key);
-            if (StringUtils.isNotEmpty(key) && StringUtils.isNotEmpty(value)) {
-                if (log.isDebugEnabled()) {
-                    String message = String.format("Searching users by claim : %s for value : %s", key, value);
-                    log.debug(message);
+        try {
+            String[] userStoreDomainNames = getDomainNames(tenantId);
+            carbonUM = (AbstractUserStoreManager) getUserStoreManager(tenantId);
+            RealmService realmService = IdentityRecoveryServiceDataHolder.getInstance().getRealmService();
+            ClaimManager claimManager = (ClaimManager) realmService.getTenantUserRealm(IdentityTenantUtil.
+                    getTenantId(tenantDomain)).getClaimManager();
+            ArrayList<org.wso2.carbon.user.core.common.User> resultedUserList = new ArrayList<>();
+
+            for (String domain : userStoreDomainNames) {
+                List<ExpressionCondition> expressionConditionList =
+                        getExpressionConditionList(claims, domain, claimManager);
+
+                if (!expressionConditionList.isEmpty()) {
+                    Condition operationalCondition = getOperationalCondition(expressionConditionList);
+                    /* Get the users list that matches with the condition
+                       limit : 2, offset : 1, sortBy : null, sortOrder : null */
+                    resultedUserList.addAll(carbonUM.getUserListWithID(operationalCondition, domain,
+                            UserCoreConstants.DEFAULT_PROFILE, 2, 1, null, null));
                 }
-                // Get the matching user list for the given claim.
-                String[] matchedUserList = getUserList(tenantId, key, value);
-                // Check for duplicates with the already retrieved users, if the matched users for the claim.
-                if (!ArrayUtils.isEmpty(matchedUserList)) {
-                    // In the first iteration resultedUserList will be empty.
-                    if (ArrayUtils.isNotEmpty(resultedUserList)) {
-                        // Remove the users from already matched list who are not in the recently matched list.
-                        resultedUserList = getCommonUserEntries(resultedUserList, matchedUserList, key, value);
-                        if (ArrayUtils.isEmpty(resultedUserList)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("No user matched for given claims");
-                            }
-                            return StringUtils.EMPTY;
-                        }
-                    } else {
-                        resultedUserList = matchedUserList;
-                    }
-                } else {
+                else {
```

Comment:
It would be much cleaner if you write the if you flip the if condition and get rid of the else block

something like

```
if (expressionConditionList.isEmpty()) {
    continue;
}

// Continue the rest of the code segment
Condition operationalCondition = getOperationalCondition(expressionConditionList);
/* Get the users list that matches with the condition
                       limit : 2, offset : 1, sortBy : null, sortOrder : null */
resultedUserList.addAll(carbonUM.getUserListWithID(operationalCondition, domain,
                            UserCoreConstants.DEFAULT_PROFILE, 2, 1, null, null));

if (resultedUserList.size() > 1) {
    // Break with exception
}
```

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/fa0dbdc33ab038eb8cae2b8d2d5fffd765503470